File: algorithms/2_medium/L/01124_longest_well_performing_interval/prefix.c
int longestWPI(const int* hours, const int len)
{
    assert(len >= 0 && len <= 10000);
    int16_t* prefix = malloc((size_t)(len + 2) * sizeof(*prefix));
    assert(prefix != NULL);
    for (int i = 0; i < len + 2; i += 1) { prefix[i] = -1; }
    int mid = len + 1;
    prefix[mid] = 0;
    int pos = mid;
    int res = 0;

    for (int i = 0; i < len; i += 1) {
        pos = hours[i] > 8 ? (pos + 1) : (pos - 1);
        if (pos > mid) {
            res = i + 1;
        } else {
            if (prefix[pos] == -1) {
                prefix[pos] = (int16_t)i;
            }

            if (prefix[pos - 1] != -1) {
                const int curr_width = i - prefix[pos - 1];
                if (curr_width > res) {
                    res = curr_width;
                }
            }
        }
    }

/* cleanup: */
    free(prefix);
    return res;
}

File: algorithms/2_medium/L/01124_longest_well_performing_interval/prefix.rs
impl Solution {
    pub fn longest_wpi(hours: Vec<i32>) -> i32 {
        let len = hours.len();
        assert!(len <= 10_000);
        let mut prefix = vec![-1_i16; len + 2];
        let mid = len + 1;
        prefix[mid] = 0;
        let mut pos = mid;
        let mut res = 0_i32;

        for (i, h) in hours.into_iter().enumerate() {
            pos = if h > 8 { pos + 1 } else { pos - 1 };

            if pos > mid {
                res = (i as i32) + 1;
            } else {
                 if prefix[pos] == -1 {
                    prefix[pos] = i as i16;
                }

                if prefix[pos - 1] != -1 {
                    let curr_width = (i as i32) - (prefix[pos - 1] as i32);
                    res = res.max(curr_width);
                }                
            }
        }

        res
    }
}
